File: crates/api/src/http/dto/response/network_status.rs
//! Response DTO for `GET /api/network/status`.
//!
//! Kept separate from the domain types, like every other `*Response`
//! in this module — the domain model never leaks into the JSON wire
//! shape.

use chrono::{DateTime, Utc};
use serde::Serialize;

use yog_core::domain::{FreshnessStatus, NetworkStatus};

/// The "Solana Live" panel payload.
///
/// Two concerns combined: the chain link (slot + RPC latency, from
/// the `network_status` singleton) and ingestion freshness (derived
/// from the most recent indexed event).
#[derive(Debug, Serialize)]
#[serde(rename_all = "camelCase")]
pub(crate) struct NetworkStatusResponse {
    /// Latest Solana slot observed by the indexer.
    ///
    /// Serialized as a string: slots are u64 and can exceed the safe
    /// integer range of a JSON number consumer.
    slot: String,

    /// Round-trip latency of the indexer's last `getSlot` call, in ms.
    rpc_latency_ms: u32,

    /// When the indexer recorded the slot above.
    observed_at: DateTime<Utc>,

    /// Ingestion freshness verdict: "live" | "delayed" | "stale".
    freshness: String,

    /// Timestamp of the most recent indexed event, if any. `null`
    /// when the database holds no events yet.
    last_event_at: Option<DateTime<Utc>>,
}

impl NetworkStatusResponse {
    /// Assemble the response from its two sources: the persisted
    /// network status and the derived freshness.
    pub(crate) fn new(
        status: NetworkStatus,
        freshness: FreshnessStatus,
        last_event_at: Option<DateTime<Utc>>,
    ) -> Self {
        Self {
            slot: status.slot.to_string(),
            rpc_latency_ms: status.rpc_latency_ms,
            observed_at: status.observed_at,
            freshness: freshness.as_str().to_string(),
            last_event_at,
        }
    }
}

File: web/src/lib/api/schema/network-status.ts
/**
 * Schema for the `GET /api/network/status` response.
 *
 * Mirrors the `NetworkStatusResponse` DTO emitted by yog-api. The
 * payload combines two concerns: the chain link (slot + RPC latency,
 * from the `network_status` singleton) and ingestion freshness
 * (derived server-side from the most recent indexed event).
 *
 * Notes on a couple of fields:
 *   - `slot` is a STRING on the wire — slots are u64 and can exceed
 *     the JS safe-integer range, so yog-api serialises them as text.
 *     Kept as a string here; the UI displays it verbatim and never
 *     does arithmetic on it.
 *   - `freshness` is a closed enum — `z.enum` rejects any value
 *     outside the three known tags, so an unexpected verdict surfaces
 *     as a validation failure rather than leaking into the UI.
 */

import * as z from "zod";

/** The three ingestion-freshness verdicts yog-api can return. */
export const FreshnessSchema = z.enum(["live", "delayed", "stale"]);

/** Freshness verdict — `"live" | "delayed" | "stale"`. */
export type Freshness = z.infer<typeof FreshnessSchema>;

/** Schema for the full network status payload. */
export const NetworkStatusSchema = z.object({
  // u64 slot, serialised as a string by yog-api (see file header).
  slot: z.string(),
  // getSlot round-trip latency measured by the indexer's reporter.
  rpcLatencyMs: z.number(),
  // When the indexer recorded the slot above.
  observedAt: z.string(),
  // Ingestion freshness verdict.
  freshness: FreshnessSchema,
  // Timestamp of the most recent indexed event; null on an empty DB.
  lastEventAt: z.string().nullable(),
});

/** Validated network status payload. */
export type NetworkStatusResponse = z.infer<typeof NetworkStatusSchema>;